examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+nx.draw_spring(cora_graph, node_size=15, node_color=subjects)
+
+degrees = [cora_graph.degree(node) for node in cora_graph]
+max_degree = max(degrees)
+average_degree = sum(degrees) / cora_graph.number_of_nodes()
+print(f"Average number of citations per paper: {round(average_degree)}.")
+print(f"Maximum number of citations per paper: {max_degree}.")
+
+"""
+### Split the dataset into stratified train and test sets
+"""
+
+data_size = papers.shape[0]
+train_ratio = 0.5
+
+train_data = []
+test_data = []
+
+groups = papers.groupby("subject")
+for _, group_data in groups:
+    random_selection = np.random.rand(len(group_data.index)) <= train_ratio
+    train_data.append(group_data[random_selection])
+    test_data.append(group_data[~random_selection])
+
+train_data = pd.concat(train_data).sample(frac=1)
+test_data = pd.concat(test_data).sample(frac=1)
+
+print("Train data shape:", train_data.shape)
+print("Test data shape:", test_data.shape)
+
+"""
+## Implement Train and Evaluate Experiment
+"""
+
+hidden_units = [32, 32]
+learning_rate = 0.01
+dropout_rate = 0.5
+num_epochs = 300
+batch_size = 256
+
+"""
+This function compiles and trains an input model using the given training data.
+"""
+
+
+def run_experiment(model, x_train, y_train):
+    # Compile the model.
+    model.compile(
+        optimizer=keras.optimizers.Adam(learning_rate),
+        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+        metrics=[keras.metrics.SparseCategoricalAccuracy(name="acc")],
+    )
+    # Create an early stopping callback.
+    early_stopping = keras.callbacks.EarlyStopping(
+        monitor="val_acc", patience=50, restore_best_weights=True
+    )
+    # Fit the model.
+    history = model.fit(
+        x=x_train,
+        y=y_train,
+        epochs=num_epochs,
+        batch_size=batch_size,
+        validation_split=0.15,
+        callbacks=[early_stopping],
+    )
+
+    return history
+
+
+"""
+This function displays the loss and accuracy curves of the model during training.
+"""
+
+
+def display_learning_curves(history):
+    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 5))
+
+    ax1.plot(history.history["loss"])
+    ax1.plot(history.history["val_loss"])
+    ax1.legend(["train", "test"], loc="upper right")
+    ax1.set_xlabel("Epochs")
+    ax1.set_ylabel("loss")
+
+    ax2.plot(history.history["acc"])
+    ax2.plot(history.history["val_acc"])
+    ax2.legend(["train", "test"], loc="upper right")
+    ax2.set_xlabel("Epochs")
+    ax2.set_ylabel("Accuracy")
+    plt.show()
+
+
+"""
+## Implement Feedforward Network (FFN) Module
+
+We will use this module in the baseline and the GNN models.
+"""
+
+
+def create_ffn(hidden_units, dropout_rate, name=None):
+    fnn_layers = []
+
+    for units in hidden_units:
+        fnn_layers.append(layers.BatchNormalization())
+        fnn_layers.append(layers.Dropout(dropout_rate))
+        fnn_layers.append(layers.Dense(units, activation=tf.nn.gelu))
```

Comment:
You sure you don't want to apply the activation after the BN layer?

Reply:
You mean before the dropout? I think it shouldn't make much difference. I tried to follow [Design Space for Graph NN](https://arxiv.org/pdf/2011.08843.pdf) paper (see Figure 1).